File: app/src/main/java/com/liweiyap/narradir/util/audio/BackgroundSoundTestMediaPlayer.java
package com.liweiyap.narradir.util.audio;

import android.content.Context;
import android.media.MediaPlayer;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.liweiyap.narradir.R;

public class BackgroundSoundTestMediaPlayer implements IMediaPlayerController
{
    public BackgroundSoundTestMediaPlayer(final @NonNull Context context)
    {
        mContext = context;
    }

    public void destroy()
    {
        mContext = null;
        free();
    }

    @Override
    public void play(final String res, final float volume, final MediaPlayer.OnCompletionListener listener)
    {
        if (mMediaPlayer != null)
        {
            mMediaPlayer.stop();
        }

        if (res == null)
        {
            return;
        }

        try
        {
            mMediaPlayer = create(res);
            if (mMediaPlayer == null)
            {
                return;
            }

            mMediaPlayer.setLooping(true);
            setVolume(volume);
            mMediaPlayer.setOnCompletionListener(listener);
            mMediaPlayer.start();
        }
        catch (IllegalStateException e)
        {
            e.printStackTrace();
        }
    }

    @Override
    public void resume()
    {
        if (mMediaPlayer == null)
        {
            return;
        }

        if (mIsPlaying)
        {
            mMediaPlayer.start();
        }
    }

    @Override
    public void pause()
    {
        if (mMediaPlayer == null)
        {
            return;
        }

        mIsPlaying = mMediaPlayer.isPlaying();
        mMediaPlayer.pause();
    }

    @Override
    public void stop()
    {
        if (mMediaPlayer == null)
        {
            return;
        }

        mMediaPlayer.setOnCompletionListener(null);
        mMediaPlayer.stop();
    }

    @Override
    public void free()
    {
        if (mMediaPlayer == null)
        {
            return;
        }

        mMediaPlayer.release();
        mMediaPlayer = null;
    }

    @Override
    public void setVolume(final float volume)
    {
        if (mMediaPlayer == null)
        {
            return;
        }

        mMediaPlayer.setVolume(volume, volume);
    }

    private @Nullable MediaPlayer create(final @NonNull String res)
    {
        if (mContext == null)
        {
            return null;
        }

        // does not look elegant but it's safe because RawRes ID integers are non-constant from Gradle version >4.
        // no need to call prepare(); create() does that for you (https://stackoverflow.com/a/59682667/12367873)
        if (res.equals(mContext.getString(R.string.backgroundsound_cards)))
        {
            return MediaPlayer.create(mContext, R.raw.backgroundcards);
        }
        else if (res.equals(mContext.getString(R.string.backgroundsound_crickets)))
        {
            return MediaPlayer.create(mContext, R.raw.backgroundcrickets);
        }
        else if (res.equals(mContext.getString(R.string.backgroundsound_fireplace)))
        {
            return MediaPlayer.create(mContext, R.raw.backgroundfireplace);
        }
        else if (res.equals(mContext.getString(R.string.backgroundsound_rain)))
        {
            return MediaPlayer.create(mContext, R.raw.backgroundrain);
        }
        else if (res.equals(mContext.getString(R.string.backgroundsound_rainforest)))
        {
            return MediaPlayer.create(mContext, R.raw.backgroundrainforest);
        }
        else if (res.equals(mContext.getString(R.string.backgroundsound_rainstorm)))
        {
            return MediaPlayer.create(mContext, R.raw.backgroundrainstorm);
        }
        else if (res.equals(mContext.getString(R.string.backgroundsound_wolves)))
        {
            return MediaPlayer.create(mContext, R.raw.backgroundwolves);
        }

        return null;
    }

    private Context mContext;

    private MediaPlayer mMediaPlayer;
    private boolean mIsPlaying;
}

File: app/src/main/java/com/liweiyap/narradir/util/audio/BackgroundSoundTestMediaPlayer.kt
package com.liweiyap.narradir.util.audio

import android.content.Context
import android.media.MediaPlayer

import com.liweiyap.narradir.R

class BackgroundSoundTestMediaPlayer(context: Context): IMediaPlayerController {
    private var mContext: Context?
    private var mMediaPlayer: MediaPlayer? = null
    private var mIsPlaying: Boolean = false

    init {
        mContext = context
    }

    fun destroy() {
        mContext = null
        free()
    }

    override fun play(res: String?, volume: Float, listener: MediaPlayer.OnCompletionListener?) {
        mMediaPlayer?.stop()

        if (res == null) {
            return
        }

        try {
            mMediaPlayer = create(res)
            if (mMediaPlayer == null) {
                return
            }

            mMediaPlayer!!.isLooping = true
            setVolume(volume)
            mMediaPlayer!!.setOnCompletionListener(listener)
            mMediaPlayer!!.start()
        }
        catch (e: IllegalStateException) {
            e.printStackTrace()
        }
    }

    override fun resume() {
        if (mIsPlaying) {
            mMediaPlayer?.start()
        }
    }

    override fun pause() {
        mMediaPlayer?.let {
            mIsPlaying = it.isPlaying
        }
        mMediaPlayer?.pause()
    }

    override fun stop() {
        mMediaPlayer?.setOnCompletionListener(null)
        mMediaPlayer?.stop()
    }

    override fun free() {
        mMediaPlayer?.release()
        mMediaPlayer = null
    }

    override fun setVolume(volume: Float) {
        mMediaPlayer?.setVolume(volume, volume)
    }

    private fun create(res: String): MediaPlayer? {
        if (mContext == null) {
            return null
        }

        // does not look elegant but it's safe because RawRes ID integers are non-constant from Gradle version >4.
        // no need to call prepare(); create() does that for you (https://stackoverflow.com/a/59682667/12367873)
        return when (res) {
            mContext!!.getString(R.string.backgroundsound_cards) -> MediaPlayer.create(mContext, R.raw.backgroundcards)
            mContext!!.getString(R.string.backgroundsound_crickets) -> MediaPlayer.create(mContext, R.raw.backgroundcrickets)
            mContext!!.getString(R.string.backgroundsound_fireplace) -> MediaPlayer.create(mContext, R.raw.backgroundfireplace)
            mContext!!.getString(R.string.backgroundsound_rain) -> MediaPlayer.create(mContext, R.raw.backgroundrain)
            mContext!!.getString(R.string.backgroundsound_rainforest) -> MediaPlayer.create(mContext, R.raw.backgroundrainforest)
            mContext!!.getString(R.string.backgroundsound_rainstorm) -> MediaPlayer.create(mContext, R.raw.backgroundrainstorm)
            mContext!!.getString(R.string.backgroundsound_wolves) -> MediaPlayer.create(mContext, R.raw.backgroundwolves)
            else -> null
        }
    }
}